Mathlib/Analysis/InnerProductSpace/TensorProduct.lean

```diff
@@ -0,0 +1,223 @@
+  · revert y
+    induction' y' using TensorProduct.induction_on with e'' f'' y₁ y₂ ih₁ ih₂
+    · simp [inner]
+    · intro x h y h'
+      rfl
+    · intro x hx y hy
+      simp_all
+  · intro x hx
+    simp_all
+
+lemma tensor_product_aux_restrict_apply (x y : E ⊗[𝕜] F)
+    (E' : Submodule 𝕜 E) (F' : Submodule 𝕜 F)
+    (hx : x ∈ LinearMap.range (TensorProduct.map E'.subtype F'.subtype))
+    (hy : y ∈ LinearMap.range (TensorProduct.map E'.subtype F'.subtype)) :
+    (inner 𝕜 x y) = (inner 𝕜 hx.choose hy.choose) := by
+  apply tensor_product_aux_restrict_apply'
+  · exact (hx.choose_spec).symm
+  · exact (hy.choose_spec).symm
+
+open scoped ComplexOrder
+
+theorem TensorProduct.inner_definite (x : E ⊗[𝕜] F) (hx : inner 𝕜 x x = 0) : x = 0 := by
+  obtain ⟨E', F', iE', iF', hz⟩ := toTensorFiniteDimensional E F x
+  rw [tensor_product_aux_restrict_apply x x E' F' hz hz] at hx
+  let y := hz.choose
+  obtain e := stdOrthonormalBasis 𝕜 E'
+  obtain f := stdOrthonormalBasis 𝕜 F'
+  have hy : y = hz.choose := rfl
+  rw [← hy] at hx
+  rw [TensorProduct.of_basis_eq_span y e.toBasis f.toBasis] at hx
+  simp only [OrthonormalBasis.coe_toBasis, TensorProduct.inner_sum, TensorProduct.inner_smul,
+    TensorProduct.sum_inner, TensorProduct.smul_inner, inner_tmul] at hx
+  simp [OrthonormalBasis.inner_eq_ite] at hx
+  simp [RCLike.mul_conj, ← Finset.sum_product'] at hx
+  rw [Finset.sum_eq_zero_iff_of_nonneg (fun _ _ => by simp)] at hx
+  simp at hx
+  have : y = 0 := by
+    rw [Basis.ext_elem_iff (e.toBasis.tensorProduct f.toBasis)]
+    simp only [hx, map_zero, Finsupp.coe_zero, Pi.zero_apply, implies_true]
+  rw [← hz.choose_spec, ← hy, this, map_zero]
+
+theorem TensorProduct.re_inner_self_nonneg (x : E ⊗[𝕜] F) :
+    0 ≤ RCLike.re (inner 𝕜 x x) := by
+  obtain ⟨E', F', iE', iF', hz⟩ := toTensorFiniteDimensional E F x
+  rw [tensor_product_aux_restrict_apply x x E' F' hz hz]
+  let y := hz.choose
+  obtain e := stdOrthonormalBasis 𝕜 E'
+  obtain f := stdOrthonormalBasis 𝕜 F'
+  have hy : y = hz.choose := rfl
+  rw [← hy]
+  rw [TensorProduct.of_basis_eq_span y e.toBasis f.toBasis]
+  simp only [OrthonormalBasis.coe_toBasis, TensorProduct.inner_sum, TensorProduct.inner_smul,
+    TensorProduct.sum_inner, TensorProduct.smul_inner, inner_tmul]
+  simp only [OrthonormalBasis.inner_eq_ite, mul_ite, mul_one, mul_zero, Finset.sum_ite_eq',
+    Finset.mem_univ, ↓reduceIte, ← Finset.sum_product', RCLike.mul_conj, map_sum]
+  apply Finset.sum_nonneg
+  intro i hi
+  rw [← RCLike.ofReal_pow, RCLike.ofReal_re]
+  exact sq_nonneg _
+
+end
+
+noncomputable instance TensorProduct.instNormedAddCommGroup : NormedAddCommGroup (E ⊗[𝕜] F) :=
+  @InnerProductSpace.Core.toNormedAddCommGroup 𝕜 (E ⊗[𝕜] F) _ _ _
+  { conj_inner_symm := fun x y => TensorProduct.conj_inner y x
+    add_left := TensorProduct.add_inner
+    smul_left := TensorProduct.smul_inner
+    definite := TensorProduct.inner_definite
+    re_inner_nonneg := TensorProduct.re_inner_self_nonneg }
```

Comment:
This instance is doing a lot of heavy lifting; it also defines:
* The topology
* The uniform space structure
* The `edist` and `nndist` and `dist` functions
* The `enorm` and `nnnorm` and `norm` functions

Can you add the theorem that says what `norm (x ⊗ₜ[𝕜] y)` is, and analogous results for at least all the `dist` and `norm`s?